src/Juvix/Core/FromFrontend.hs

```diff
@@ -0,0 +1,260 @@
+          <*> transformTermHR b
+
+makeList :: [HR.Term primTy primVal] -> HR.Term primTy primVal
+makeList =
+  -- TODO: make lists a builtin syntax form in core (translated to
+  -- e.g. michelson lists)
+  _
+
+-- | translate (1,2,3,4) to (1,(2,(3,4)))
+makeTuple :: [HR.Term primTy primVal] -> HR.Term primTy primVal
+makeTuple [] = _ -- TODO: add unit to core, probably
+makeTuple [t] = t
+makeTuple (t:ts) = IR.Pair t (makeTuple ts)
+
+makeRecord :: [(NameSymbol.T, HR.Term primTy primVal)]
+           -> HR.Term primTy primVal
+makeRecord =
+  -- TODO:
+  -- 1) an extended version of core with record literals to begin with
+  -- 2) translation after typechecking of these to an application of the right
+  --    constructor to the fields in the originally declared order
+  _
+
+toElim :: FE.Expression -- ^ the original expression
+       -> HR.Term primTy primVal
+       -> Env primTy primVal (HR.Elim primTy primVal)
+toElim _ (HR.Elim e) = pure e
+toElim e _           = throwFF $ NotAnElim e
+  -- TODO put an annotation with metas for the usage/type
+
+transformUsage :: FE.Expression -> Env primTy primVal Usage.T
+transformUsage (FE.Constant (FE.Number (FE.Integer' i))) | i >= 0 =
+  pure $ Usage.SNat $ fromInteger i
+transformUsage e = throwFF $ NotAUsage e
+-- FIXME ω
+
+transformGUsage :: FE.Expression -> Env primTy primVal IR.GlobalUsage
+transformGUsage (FE.Constant (FE.Number (FE.Integer' 0))) = pure IR.GZero
+transformGUsage e = throwFF $ NotAGUsage e
+-- FIXME ω
+
+
+transformSig :: FE.Final Ctx.Definition
+             -> Env primTy primVal (CoreSigHR primTy primVal)
+transformSig def@(Ctx.Def π msig _ _) =
+  -- why two usages?
+  transformValSig def π msig
+transformSig def@(Ctx.Record _ msig) = transformValSig def Nothing msig
+transformSig (Ctx.TypeDeclar typ) = DataSig <$> transformType typ
+transformSig (Ctx.Unknown sig) =
+  throwFF $ UnknownUnsupported $ FE.signatureName <$> sig
+transformSig Ctx.CurrentNameSpace = _ -- TODO ???
+
+transformType :: forall primTy primVal.
+  FE.Type -> Env primTy primVal (HR.Term primTy primVal)
+transformType (FE.Typ {typeForm = FE.Arrowed {dataArrow}}) =
+  transformTermHR dataArrow
+transformType typ@(FE.Typ {typeForm = FE.NonArrowed {}}) =
+  throwFF $ SigRequired $ Ctx.TypeDeclar typ
+
+transformValSig :: FE.Final Ctx.Definition
+                -> Maybe Usage.T -> Maybe FE.Signature
+                -> Env primTy primVal (CoreSigHR primTy primVal)
+transformValSig _ π (Just (FE.Sig _ π' ty cons))
+  | null cons = ValSig _π <$> transformTermHR ty
+  | otherwise = throwFF $ ConstraintsUnsupported cons
+transformValSig def π Nothing = throwFF $ SigRequired def
+
+transformDef :: FE.Final Ctx.Definition
+             -> Env primTy primVal [IR.Global primTy primVal]
+transformDef (Ctx.Def π _ def _)  = _
+transformDef (Ctx.Record rec _)   = _
+transformDef (Ctx.TypeDeclar typ) = _
+transformDef (Ctx.Unknown _)      = pure []
+transformDef Ctx.CurrentNameSpace = pure [] -- FIXME ???
```

Comment:
these get called in `transformInner` in the other passes.

It seems here `transformInner` is not a function, so this function is uneeded? 

I'm not really sure how this is all called.

Reply:
The full thing when it all fits together will split the frontend context into dependency groups, and then for each group:

- get its declared type via `transformSig`, check all of them are actually types
- temporarily add them all to the core context as `GAbstract`s with the declared types
- transform & check the definitions with this temporary context (i.e., knowing the types of everything in the group, but not being able to reduce them)
- finally, extend the real core context with the actual definitions, which will be available to later groups